File: fore-adapters/src/main/java/co/early/fore/adapters/immutable/DiffableImp.java
package co.early.fore.adapters.immutable;

import androidx.recyclerview.widget.DiffUtil;

import java.util.ArrayList;
import java.util.List;

import co.early.fore.core.Affirm;
import co.early.fore.core.WorkMode;
import co.early.fore.core.logging.Logger;
import co.early.fore.core.observer.ObservableImp;
import co.early.fore.core.threading.AsyncBuilder;
import co.early.fore.core.time.SystemTimeWrapper;

public class DiffableImp<T extends DeepCopyable<T> & DiffComparator<T>> extends ObservableImp implements Diffable {

    private static final String LOG_TAG = DiffableImp.class.getSimpleName();

    private final SystemTimeWrapper systemTimeWrapper;
    private final WorkMode workMode;
    private final Logger logger;
    private DiffSpec latestDiffSpec = createFullDiffSpec();
    private List<T> currentList = new ArrayList<T>();
    private List<T> currentListMutableCopy = new ArrayList<T>();

    public DiffableImp(SystemTimeWrapper systemTimeWrapper, WorkMode workMode, Logger logger) {
        super(workMode, logger);
        this.systemTimeWrapper = Affirm.notNull(systemTimeWrapper);
        this.workMode = workMode;
        this.logger = logger;
    }

    public T getItem(int index) {
        return currentList.get(index);
    }

    public List<T> getListCopy() {
        return currentListMutableCopy;
    }

    public int size() {
        return currentList.size();
    }

    @SuppressWarnings("unchecked")
    public void updateList(List<T> newList) {

        new AsyncBuilder<Input, Result>(workMode)
                .doInBackground(input -> doWork(input[0]))
                .onPostExecute(this::updateState)
                .execute(new Input(currentList, newList));
    }

    private Result doWork(Input input) {

        // work out the differences in the lists
        DiffUtil.DiffResult diffResult = new DiffCalculator<T>().createDiffResult(input.oldList, input.newList);

        //create a mutable copy of the new list, ready for when client wants to change it
        List<T> newListCopy = new ArrayList<>(input.newList.size());
        for (T item : input.newList) {
            newListCopy.add(item.deepCopy());
        }

        //return to the UI thread
        return new Result(input.newList, newListCopy, new DiffSpec(diffResult, systemTimeWrapper));
    }

    private void updateState(Result result) {
        currentList = result.newList;
        currentListMutableCopy = result.newListCopy;
        latestDiffSpec = result.diffSpec;
        logger.i(LOG_TAG, "list updated, thread:" + Thread.currentThread().getId());
        notifyObservers();
    }

    /**
     * If the DiffResult is old, then we assume that whatever changes
     * were made to the list last time were never picked up by a
     * recyclerView (maybe because the list was not visible at the time).
     * In this case we clear the DiffResult and create a fresh one with a
     * full diff spec.
     *
     * @return the latest DiffResult for the list
     */
    @Override
    public DiffSpec getAndClearLatestDiffSpec(long maxAgeMs) {

        DiffSpec latestDiffSpecAvailable = latestDiffSpec;
        DiffSpec fullDiffSpec = createFullDiffSpec();

        latestDiffSpec = fullDiffSpec;

        if ((systemTimeWrapper.currentTimeMillis() - latestDiffSpecAvailable.timeStamp) < maxAgeMs) {
            return latestDiffSpecAvailable;
        } else {
            return fullDiffSpec;
        }
    }

    private DiffSpec createFullDiffSpec() {
        return new DiffSpec(null, systemTimeWrapper);
    }

    private class Input {

        public final List<T> oldList;
        public final List<T> newList;

        public Input(List<T> oldList, List<T> newList) {
            this.oldList = Affirm.notNull(oldList);
            this.newList = Affirm.notNull(newList);
        }
    }

    private class Result {

        public final List<T> newList;
        public final List<T> newListCopy;
        public final DiffSpec diffSpec;

        public Result(List<T> newList, List<T> newListCopy, DiffSpec diffSpec) {
            this.newList = Affirm.notNull(newList);
            this.newListCopy = Affirm.notNull(newListCopy);
            this.diffSpec = Affirm.notNull(diffSpec);
        }
    }
}

File: fore-adapters-kt/src/main/java/co/early/fore/kt/adapters/DiffableImp.kt
package co.early.fore.kt.adapters

import co.early.fore.adapters.DiffCalculator
import co.early.fore.adapters.DiffComparator
import co.early.fore.adapters.DiffSpec
import co.early.fore.adapters.Diffable
import co.early.fore.core.WorkMode
import co.early.fore.core.observer.Observable
import co.early.fore.core.time.SystemTimeWrapper
import co.early.fore.kt.core.coroutine.awaitDefault
import co.early.fore.kt.core.coroutine.launchMain
import co.early.fore.kt.core.delegate.ForeDelegateHolder
import co.early.fore.kt.core.logging.Logger
import co.early.fore.kt.core.observer.ObservableImp

class DiffableImp<T>(
        private val systemTimeWrapper: SystemTimeWrapper? = null,
        private val workMode: WorkMode? = null,
        private val logger: Logger? = null
) : Diffable, Observable by ObservableImp(workMode, logger)
        where T : DeepCopyable<T>, T : DiffComparator<T> {

    private var latestDiffSpec: DiffSpec? = createFullDiffSpec()
    private var currentList = listOf<T>()
    private var currentListMutableCopy = mutableListOf<T>()

    fun getItem(index: Int): T {
        return currentList[index]
    }

    fun getListCopy(): MutableList<T> {
        return currentListMutableCopy
    }

    fun size(): Int = currentList.size

    fun updateList(newList: List<T>) {

        launchMain(workMode) {

            val result = awaitDefault(workMode) {

                // work out the differences in the lists
                val diffResult = DiffCalculator<T>().createDiffResult(currentList, newList)

                //create a mutable copy of the new list, ready for when client wants to change it
                val newListCopy = newList.map { it.deepCopy() }

                //return to the UI thread
                Triple(newList, newListCopy.toMutableList(), DiffSpec(diffResult, ForeDelegateHolder.getSystemTimeWrapper(systemTimeWrapper)))
            }

            currentList = result.first
            currentListMutableCopy = result.second
            latestDiffSpec = result.third
            ForeDelegateHolder.getLogger(logger).i("list updated, thread:" + Thread.currentThread().id)
            notifyObservers()
        }

    }

    /**
     * If the DiffResult is old, then we assume that whatever changes
     * were made to the list last time were never picked up by a
     * recyclerView (maybe because the list was not visible at the time).
     * In this case we clear the DiffResult and create a fresh one with a
     * full diff spec.
     *
     * @return the latest DiffResult for the list
     */
    override fun getAndClearLatestDiffSpec(maxAgeMs: Long): DiffSpec {

        val latestDiffSpecAvailable = latestDiffSpec
        val fullDiffSpec = createFullDiffSpec()

        latestDiffSpec = fullDiffSpec

        return if (ForeDelegateHolder.getSystemTimeWrapper(systemTimeWrapper).currentTimeMillis() - latestDiffSpecAvailable!!.timeStamp < maxAgeMs) {
            latestDiffSpecAvailable
        } else {
            fullDiffSpec
        }
    }

    private fun createFullDiffSpec(): DiffSpec {
        return DiffSpec(null, ForeDelegateHolder.getSystemTimeWrapper(systemTimeWrapper))
    }
}
